src/Microsoft.Identity.Web/AzureADB2COpenIDConnectEventHandlers.cs

```diff
@@ -10,103 +10,103 @@
-
-                context.Response.Redirect($"{context.Request.PathBase}/MicrosoftIdentity/Account/Error");
-            }
-
-            return Task.CompletedTask;
-        }
-
-        private string BuildIssuerAddress(RedirectContext context, string? defaultUserFlow, string userFlow)
-        {
-            if (!_userFlowToIssuerAddress.TryGetValue(userFlow, out var issuerAddress))
-            {
-                _userFlowToIssuerAddress[userFlow] = context.ProtocolMessage.IssuerAddress.ToLowerInvariant()
-                    .Replace($"/{defaultUserFlow?.ToLowerInvariant()}/", $"/{userFlow.ToLowerInvariant()}/");
-            }
-
-            return _userFlowToIssuerAddress[userFlow];
-        }
-    }
+	internal class AzureADB2COpenIDConnectEventHandlers
+	{
+		private readonly ILoginErrorAccessor _errorAccessor;
+
+		private readonly IDictionary<string, string> _userFlowToIssuerAddress =
+			new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+		public AzureADB2COpenIDConnectEventHandlers(
+			string schemeName,
+			MicrosoftIdentityOptions options,
+			ILoginErrorAccessor errorAccessor)
+		{
+			SchemeName = schemeName;
+			Options = options;
+			_errorAccessor = errorAccessor;
+		}
+
+		public string SchemeName { get; }
+
+		public MicrosoftIdentityOptions Options { get; }
+
+		public Task OnRedirectToIdentityProvider(RedirectContext context)
+		{
+			var defaultUserFlow = Options.DefaultUserFlow;
+			if (context.Properties.Items.TryGetValue(OidcConstants.PolicyKey, out var userFlow) &&
+				!string.IsNullOrEmpty(userFlow) &&
+				!string.Equals(userFlow, defaultUserFlow, StringComparison.OrdinalIgnoreCase))
+			{
+				context.ProtocolMessage.IssuerAddress = BuildIssuerAddress(context, defaultUserFlow, userFlow);
+				context.Properties.Items.Remove(OidcConstants.PolicyKey);
+
+				if (!Options.HasClientCredentials)
+				{
+					context.ProtocolMessage.ResponseType = OpenIdConnectResponseType.IdToken;
+				}
+				else
+				{
+					if (Options.IsB2C)
+					{
+						context.ProtocolMessage.ResponseType = OpenIdConnectResponseType.CodeIdToken;
+					}
+					else
+					{
+						context.ProtocolMessage.ResponseType = OpenIdConnectResponseType.Code;
+					}
+				}
+			}
+
+			return Task.CompletedTask;
+		}
+
+		public Task OnRemoteFailure(RemoteFailureContext context)
+		{
+			context.HandleResponse();
+
+			// Handle the error code that Azure Active Directory B2C throws when trying to reset a password from the login page
+			// because password reset is not supported by a "sign-up or sign-in user flow".
+			// Below is a sample error message:
+			// 'access_denied', error_description: 'AADB2C90118: The user has forgotten their password.
+			// Correlation ID: f99deff4-f43b-43cc-b4e7-36141dbaf0a0
+			// Timestamp: 2018-03-05 02:49:35Z
+			// ', error_uri: 'error_uri is null'.
+			if (context.Failure is OpenIdConnectProtocolException && context.Failure.Message.Contains(ErrorCodes.B2CForgottenPassword))
```

Comment:
context.Failure is OpenIdConnectProtocolException can be evaluated once rather than re-evaluated in multiple if statements.